ClientRuntimes/Python/msrestazure/msrestazure/azure_active_directory.py

```diff
@@ -216,7 +228,70 @@ def clear_cached_token(self):
             raise_with_traceback(KeyError, "Unable to clear token.")
 
 
-class UserPassCredentials(OAuthTokenAuthentication, AADMixin):
+class AADRefreshMixin(object):
+    """
+    Additional token refresh logic
+    """
+
+    def refresh_session(self):
+        """Return updated session if token has expired, attempts to
+        refresh using newly acquired token.
+
+        :rtype: requests.Session.
+        """
+        if self.token.get('refresh_token'):
+            try:
+                return self.signed_session()
+            except Expired:
+                pass
+        self.set_token()
+        return self.signed_session()
+
+
+class AADTokenCredentials(AADMixin):
+    """
+    Credentials objects for AAD token retrieved through external process
+    e.g. Python ADAL lib.
+
+    Optional kwargs may include:
+    - china (bool): Configure auth for China-based service,
```

Comment:
As of today that is a manual operation. The base uri defaults to the value in the swagger spec. However, the user can provide a different uri. The credentials object will take the right auth uri from the environment.
All they have to do is:
```
var msrestAzure = require('ms-rest-azure');
var chinaCloud = msrestAzure.AzureEnvironment.AzureChina;
var StorageClient = require('azure-arm-storage');
var options = {environment: chinaCloud};
msRestAzure.loginWithUsernamePassword('username', 'password', options, function (err, credentials) {
  var client = new StorageClient(credentials, chinacloud.resourceManagerEndpointUrl);
  client.storageAccounts.list('rgname', function(err, result, req, res) {
    if (err) return console.log(err);
    return console.log(result);
  });
});
```